EC2 update_auto_scaling_group Does Not Update Version when New Version Is Submitted

I'm trying to use [update_auto_scaling_group](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/autoscaling/client/update_auto_scaling_group.html) to update my ASGs when there is a new AMI published to the SSM Parameter. However, when writing the unit tests for this I discovered the version does not update on the ASG when using `update_auto_scaling_group` in `moto`. When run against AWS, it updates no problem. I am using a mixed instance policy, but it fails just the same using a regular launch template.

I've included a test script and my output to help with reproduction of the bug.

| Package | Version |
| --- | --- |
| moto | 5.1.14 |
| boto3 | 1.40.52 |
| Python | 3.11 |

Thank you!

## Test script

Run using `python -m unittest test.py`.

```python
"""Tests for moto github issue"""

import os
import unittest
import json

import boto3
from moto import mock_aws


def promote_launch_template_on_asg(asg_client, asg_name: str, launch_template: dict) -> dict:
    """Update mixed instance policy launch template to latest version."""
    asg_client.update_auto_scaling_group(
        AutoScalingGroupName=asg_name,
        MixedInstancesPolicy={
            "LaunchTemplate": {
                "LaunchTemplateSpecification": {
                    "LaunchTemplateName": launch_template["LaunchTemplateName"],
                    "Version": str(launch_template["VersionNumber"]),
                }
            }
        },
    )
    return asg_client.describe_auto_scaling_groups(AutoScalingGroupNames=[asg_name])["AutoScalingGroups"][0]


@mock_aws
class TestMain(unittest.TestCase):
    def setUp(self):
        os.environ["AWS_REGION"] = "us-east-1"
        os.environ["AWS_ACCESS_KEY_ID"] = "testing"
        os.environ["AWS_SECRET_ACCESS_KEY"] = "testing"
        os.environ["AWS_SECURITY_TOKEN"] = "testing"
        os.environ["AWS_SESSION_TOKEN"] = "testing"
        os.environ["AWS_DEFAULT_REGION"] = "us-east-1"
        os.environ["ACCOUNT_NUMBER"] = "123456789012"
        os.environ["ACCOUNT_ID"] = "123456789012"
        self.mock_ec2_client_east = boto3.client("ec2", region_name="us-east-1")
        self.mock_asg_client_east = boto3.client("autoscaling", region_name="us-east-1")
        self.mock_cfn_client_east = boto3.client("cloudformation", region_name="us-east-1")

        # Create VPC for ASG
        vpc = self.mock_ec2_client_east.create_vpc(CidrBlock="10.0.0.0/16")
        vpc_id = vpc["Vpc"]["VpcId"]
        # create Subnet for ASG
        subnet = self.mock_ec2_client_east.create_subnet(
            CidrBlock="10.0.0.0/24",
            AvailabilityZone="us-east-1a",
            VpcId=vpc_id,
        )
        subnet_id = subnet["Subnet"]["SubnetId"]
        # Create Launch Template
        self.mock_ec2_client_east.create_launch_template(
            LaunchTemplateName="test-lt",
            LaunchTemplateData={"ImageId": "ami-67890abcdeEXAMPLE"},
        )
        # Create ASG CFN Stack
        self.mock_cfn_client_east.create_stack(
            StackName="asg_stack",
            TemplateBody=json.dumps(
                {
                    "AWSTemplateFormatVersion": "2010-09-09",
                    "Resources": {
                        "SomeASG": {
                            "Type": "AWS::AutoScaling::AutoScalingGroup",
                            "Properties": {
                                "AutoScalingGroupName": "test-asg",
                                "MinSize": "1",
                                "MaxSize": "1",
                                "VPCZoneIdentifier": [subnet_id],
                                "MixedInstancesPolicy": {
                                    "LaunchTemplate": {
                                        "LaunchTemplateSpecification": {
                                            "LaunchTemplateName": "test-lt",
                                            "Version": "1",
                                        }
                                    }
                                },
                            },
                        }
                    },
                }
            ),
        )

    def test_promote_launch_template_on_asg(self):
        """Test promote_launch_template_on_asg()."""
        # Describe our launch template
        launch_template = self.mock_ec2_client_east.describe_launch_templates(LaunchTemplateNames=["test-lt"])[
            "LaunchTemplates"
        ][0]
        # Create a new launch template version with our new AMI ID
        new_launch_template = self.mock_ec2_client_east.create_launch_template_version(
            LaunchTemplateId=launch_template["LaunchTemplateId"],
            SourceVersion=str(launch_template["LatestVersionNumber"]),
            LaunchTemplateData={
                "ImageId": "ami-12345abcdeEXAMPLE",
            },
        )["LaunchTemplateVersion"]
        # New launch template version is 2.
        self.assertEqual(new_launch_template["VersionNumber"], 2)
        # Update ASG to use the new launch template version.
        asg = promote_launch_template_on_asg(self.mock_asg_client_east, "test-asg", new_launch_template)
        # Print the version number to STDOUT. It shows version is 1 even though we updated it to 2.
        print(asg["MixedInstancesPolicy"]["LaunchTemplate"]["LaunchTemplateSpecification"])
        self.assertEqual(asg["MixedInstancesPolicy"]["LaunchTemplate"]["LaunchTemplateSpecification"]["Version"], "2")

```

## Test output

```
{'LaunchTemplateId': 'lt-e6d7775a8af1b891c', 'LaunchTemplateName': 'test-lt', 'Version': '1'}
F
======================================================================
FAIL: test_promote_launch_template_on_asg (test.TestMain.test_promote_launch_template_on_asg)
Test promote_launch_template_on_asg().
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/snip/testing/venv/lib/python3.11/site-packages/moto/core/models.py", line 122, in wrapper
    result = func(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^
  File "/snip/testing/tests/test.py", line 106, in test_promote_launch_template_on_asg
    self.assertEqual(asg["MixedInstancesPolicy"]["LaunchTemplate"]["LaunchTemplateSpecification"]["Version"], "2")
AssertionError: '1' != '2'
- 1
+ 2


----------------------------------------------------------------------
Ran 1 test in 0.714s

FAILED (failures=1)
```

--- a/moto/autoscaling/models.py
+++ b/moto/autoscaling/models.py
@@ -935,6 +935,7 @@ class FakeAutoScalingGroup(CloudFormationModel):
         health_check_period: int,
         health_check_type: str,
         new_instances_protected_from_scale_in: Optional[bool] = None,
+        mixed_instances_policy: Optional[dict[str, Any]] = None,
     ) -> None:
         self._set_azs_and_vpcs(availability_zones, vpc_zone_identifier, update=True)
 
@@ -949,8 +950,11 @@ class FakeAutoScalingGroup(CloudFormationModel):
             if max_size is not None and max_size < len(self.instance_states):
                 desired_capacity = max_size
 
+        self.mixed_instances_policy = mixed_instances_policy
         self._set_launch_configuration(
-            launch_config_name, launch_template, mixed_instances_policy=None
+            launch_config_name,
+            launch_template,
+            mixed_instances_policy=mixed_instances_policy,
         )
 
         if health_check_period is not None:
@@ -1424,11 +1428,11 @@ class AutoScalingBackend(BaseBackend):
         health_check_period: int,
         health_check_type: str,
         new_instances_protected_from_scale_in: Optional[bool] = None,
+        mixed_instances_policy: Optional[dict[str, Any]] = None,
     ) -> FakeAutoScalingGroup:
         """
         The parameter DefaultCooldown, PlacementGroup, TerminationPolicies are not yet implemented
         """
-        # TODO: Add MixedInstancesPolicy once implemented.
         # Verify only a single launch config-like parameter is provided.
         if launch_config_name and launch_template:
             raise ValidationError(
@@ -1450,6 +1454,7 @@ class AutoScalingBackend(BaseBackend):
             health_check_period=health_check_period,
             health_check_type=health_check_type,
             new_instances_protected_from_scale_in=new_instances_protected_from_scale_in,
+            mixed_instances_policy=mixed_instances_policy,
         )
         return group
 

--- a/moto/autoscaling/responses.py
+++ b/moto/autoscaling/responses.py
@@ -252,6 +252,7 @@ class AutoScalingResponse(BaseResponse):
             new_instances_protected_from_scale_in=self._get_bool_param(
                 "NewInstancesProtectedFromScaleIn", None
             ),
+            mixed_instances_policy=self._get_param("MixedInstancesPolicy"),
         )
         return EmptyResult()
 

--- a/tests/test_autoscaling/__init__.py
+++ b/tests/test_autoscaling/__init__.py
@@ -14,6 +14,7 @@ pytest.register_assert_rewrite("tests.test_ec2.helpers")
 def autoscaling_aws_verified(
     create_auto_scaling_group: bool = False,
     get_group_name: bool = False,
+    wait_for_instance_termination: bool = True,
 ):
     def inner(func):
         @wraps(func)
@@ -24,6 +25,7 @@ def autoscaling_aws_verified(
                 return _invoke_func(
                     create_auto_scaling_group=create_auto_scaling_group,
                     get_group_name=get_group_name,
+                    wait_for_instance_termination=wait_for_instance_termination,
                     func=func,
                     kwargs=kwargs,
                 )
@@ -36,6 +38,7 @@ def autoscaling_aws_verified(
 def _invoke_func(
     create_auto_scaling_group: bool,
     get_group_name: bool,
+    wait_for_instance_termination: bool,
     func,
     kwargs,
 ):
@@ -71,11 +74,11 @@ def _invoke_func(
                     ForceDelete=True,
                 )
 
-                # Wait until instances are terminated
-                ec2_client = boto3.client("ec2", "us-east-1")
-                waiter = ec2_client.get_waiter("instance_terminated")
-                waiter.wait(
-                    # Delay to get around https://github.com/boto/boto3/issues/176
-                    WaiterConfig={"Delay": 10},
-                    InstanceIds=[inst["InstanceId"] for inst in group["Instances"]],
-                )
+                if wait_for_instance_termination:
+                    ec2_client = boto3.client("ec2", "us-east-1")
+                    waiter = ec2_client.get_waiter("instance_terminated")
+                    waiter.wait(
+                        # Delay to get around https://github.com/boto/boto3/issues/176
+                        WaiterConfig={"Delay": 10},
+                        InstanceIds=[inst["InstanceId"] for inst in group["Instances"]],
+                    )

--- a/tests/test_autoscaling/test_autoscaling.py
+++ b/tests/test_autoscaling/test_autoscaling.py
@@ -146,7 +146,8 @@ def test_create_autoscaling_group_from_invalid_instance_id():
     launch_template_data={"ImageId": get_valid_ami(), "InstanceType": "t2.medium"},
     return_launch_template_details=True,
 )
-@autoscaling_aws_verified(get_group_name=True)
+# Instances are created in the Subnet - wait until the instances are terminated at the end, otherwise subsequent Subnet deletion fails (because of dangling resources)
+@autoscaling_aws_verified(get_group_name=True, wait_for_instance_termination=True)
 @pytest.mark.aws_verified
 def test_create_autoscaling_group_from_template(
     autoscaling_client=None,
@@ -1222,6 +1223,66 @@ def test_create_auto_scaling_group_with_mixed_instances_policy_overrides(
             ],
         },
     }
+
+
+@ec2_aws_verified(
+    create_launch_template=True,
+    launch_template_data={"ImageId": get_valid_ami(), "InstanceType": "t2.medium"},
+)
+@autoscaling_aws_verified(get_group_name=True, wait_for_instance_termination=False)
+@pytest.mark.aws_verified
+def test_update_mixed_instances_policy(
+    autoscaling_client=None,
+    autoscaling_group_name=None,
+    ec2_client=None,
+    launch_template_name=None,
+):
+    autoscaling_client.create_auto_scaling_group(
+        MixedInstancesPolicy={
+            "LaunchTemplate": {
+                "LaunchTemplateSpecification": {
+                    "LaunchTemplateName": launch_template_name,
+                    "Version": "$Default",
+                },
+            }
+        },
+        AutoScalingGroupName=autoscaling_group_name,
+        MinSize=2,
+        MaxSize=2,
+        AvailabilityZones=["us-east-1a"],
+    )
+
+    new_launch_template = ec2_client.create_launch_template_version(
+        LaunchTemplateName=launch_template_name,
+        SourceVersion="$Default",
+        LaunchTemplateData={
+            "InstanceType": "t2.large",
+        },
+    )["LaunchTemplateVersion"]
+    new_version = new_launch_template["VersionNumber"]
+
+    # Update MixedInstancesPolicy
+    autoscaling_client.update_auto_scaling_group(
+        AutoScalingGroupName=autoscaling_group_name,
+        MixedInstancesPolicy={
+            "LaunchTemplate": {
+                "LaunchTemplateSpecification": {
+                    "LaunchTemplateName": launch_template_name,
+                    "Version": str(new_launch_template["VersionNumber"]),
+                }
+            }
+        },
+    )
+
+    # Assert the MixedInstancesPolicy has an updated LaunchTemplateVersion
+    response = autoscaling_client.describe_auto_scaling_groups(
+        AutoScalingGroupNames=[autoscaling_group_name]
+    )
+    group = response["AutoScalingGroups"][0]
+    template = group["MixedInstancesPolicy"]["LaunchTemplate"]
+    template_spec = template["LaunchTemplateSpecification"]
+    assert template_spec["LaunchTemplateName"] == launch_template_name
+    assert template_spec["Version"] == str(new_version)
 
 
 @mock_aws

